proposals/lip-0052.md

```diff
@@ -362,712 +379,1832 @@ The module provides the following commands to modify the NFT store.
+                    "chainID",
+                    "supportedNftIDArray"
+                ],
+                "properties": {
+                    "chainID": {
+                        "dataType": "bytes",
+                        "length": CHAIN_ID_LENGTH,
+                        "fieldNumber": 1
+                    },
+                    "supportedNftIDArray": {
+                        "type": "array",
+                        "fieldNumber": 2,
+                        "items": {
+                            "dataType": "bytes",
+                            "length": NFT_ID_LENGTH
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
 ```
 
+#### Genesis State Initialization
+
+During the genesis state initialization stage, the following steps are executed. If any step fails, the block is discarded and has no further effect.
+
+Let `genesisBlockAssetBytes` be the `data` bytes included in the block assets for the NFT module and let `genesisBlockAssetObject` be the deserialization of `genesisBlockAssetBytes` according to the `genesisNFTStoreSchema` schema, given above.
+
+- Initial checks on the properties of `genesisBlockAssetObject`:
+
+  - Across all elements of the `NFTSubstore` array, all values given for `nftID` must be unique.
+  - For all elements of the `NFTSubstore` array, values given for `owner` must have either length `ADDRESS_ID_LENGTH` bytes (representing a user address) or `CHAIN_ID_LENGTH` bytes (representing a chain ID).
+  - The `NFTSubstore` must be in lexicographical order of `owner`.
+    - For a given `owner`, the entries must be in ascending order of `nftID`.
+    - The `attributesArray` must be in lexicographic order of `module` name.
+  - Across all elements of the `userSubstore` array, the pairs `address`, `nftID` must be unique: each pair appears at most once, although a given `address` can appear multiple times with different unique `nftID`.
+  - The `userSubstore` array must be in lexicographical order of `address`. For a given `address`, the entries must be in lexicographic order of `nftID`.
+  - Across all elements of the `escrowSubstore` array, the pairs `escrowedChainID`, `nftID` must be unique: each pair appears at most once, although a given `escrowedChainID` can appear multiple times with different unique `nftID`.
+  - The `supportedNFTsSubstore` array must be ordered lexicographically by `chainID`. For each entry of this array, the `supportedNftIDArray` should be in lexicographical order.
+
+- For each entry `NFTEntry` in `genesisBlockAssetObject.NFTSubstore`, create an entry in the NFT substore with:
+
+  ```python
+  storeKey = NFTEntry.nftID
+  storeValue = encode(
+      schema = NFTStoreSchema,
+      object = {
+          "owner": NFTEntry.owner,
+          "attributesArray": NFTEntry.attributesArray
+      }
+  )
+  ```
+
+  Furthermore, if `NFTEntry.owner` has length `ADDRESS_ID_LENGTH` bytes, create an entry in the user substore with:
```

Comment:
We do not support a genesis state with locked NFTs. This is fine for me, but in principle we could quite easily allow for that.

Reply:
Right. We can also discuss this briefly live.